app/recurringjob/volume.go

```diff
@@ -0,0 +1,651 @@
+	specLabels := map[string]string{}
+	if recurringJob.Spec.Labels != nil {
+		specLabels = recurringJob.Spec.Labels
+	}
+	specLabels[types.RecurringJobLabel] = recurringJob.Name
+
+	specLabelsJSON, err := json.Marshal(specLabels)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get JSON encoding for labels")
+	}
+
+	snapshotName := sliceStringSafely(types.GetCronJobNameForRecurringJob(job.name), 0, 8) + "-" + util.UUID()
+
+	logger := job.logger.WithFields(logrus.Fields{
+		// job-specific fields
+		"job":            job.name,
+		"task":           job.task,
+		"retain":         job.retain,
+		"parameters":     job.parameters,
+		"executionCount": job.executionCount,
+		// volume-specific fields
+		"concurrent":   recurringJob.Spec.Concurrency,
+		"groups":       strings.Join(groups, ","),
+		"volumeName":   volumeName,
+		"snapshotName": snapshotName,
+		"specLabels":   string(specLabelsJSON),
+	})
+
+	newJob := &VolumeJob{
+		Job:          job,
+		logger:       logger,
+		concurrent:   recurringJob.Spec.Concurrency,
+		groups:       groups,
+		volumeName:   volumeName,
+		snapshotName: snapshotName,
+		specLabels:   specLabels,
+	}
+	return newJob, nil
+}
+
+func (job *VolumeJob) run() (err error) {
+	job.logger.Info("Starting volume job")
+
+	volumeAPI := job.api.Volume
+	volumeName := job.volumeName
+	volume, err := volumeAPI.ById(volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "could not get volume %v", volumeName)
+	}
+
+	if len(volume.Controllers) > 1 {
+		return fmt.Errorf("cannot run job for volume %v that is using %v engines", volume.Name, len(volume.Controllers))
+	}
+
+	if volume.State != string(longhorn.VolumeStateAttached) && volume.State != string(longhorn.VolumeStateDetached) {
+		return fmt.Errorf("volume %v is in an invalid state for recurring job: %v. Volume must be in state Attached or Detached", volumeName, volume.State)
+	}
+
+	// only recurring job types `snapshot` and `backup` need to check if old snapshots can be deleted or not before creating
+	switch job.task {
+	case longhorn.RecurringJobTypeSnapshot, longhorn.RecurringJobTypeBackup:
+		if err := job.doSnapshotCleanup(false); err != nil {
+			return err
+		}
+	}
+
+	switch job.task {
+	case longhorn.RecurringJobTypeFilesystemTrim:
+		job.logger.Infof("Running recurring filesystem trim for volume %v", volumeName)
+		return job.doRecurringFilesystemTrim(volume)
+
+	case longhorn.RecurringJobTypeBackup, longhorn.RecurringJobTypeBackupForceCreate:
+		job.logger.Infof("Running recurring backup for volume %v", volumeName)
+		return job.doRecurringBackup()
+
+	default:
+		job.logger.Infof("Running recurring snapshot for volume %v", volumeName)
+		return job.doRecurringSnapshot()
+	}
+}
+
+func (job *VolumeJob) doRecurringSnapshot() (err error) {
+	defer func() {
+		err = errors.Wrap(err, "failed recurring snapshot")
+		if err == nil {
+			job.logger.Info("Finished recurring snapshot")
+		}
```

Comment:
Can we log the error here?

Reply:
Isn't this already being logged by the caller?
```
	err = volumeJob.run()
	if err != nil {
		volumeJob.logger.WithError(err).Error("Failed to run volume job")
		return err
	}
```

Reply:
Oh, yes. The error and info are logged in different places..